packages/core/src/indexing-store/historical.ts

```diff
@@ -142,51 +147,73 @@ export const getHistoricalStore = ({
-            const updateObject = data({ current });
-            // Here, `latestRow` is already encoded, so we need to exclude it from `encodeRow`.
-            const updateRow = {
-              id: latestRow.id,
-              ...encodeRow(updateObject, table, kind),
-            };
-
-            const row = await tx
-              .withSchema(namespaceInfo.userNamespace)
-              .updateTable(tableName)
-              .set(updateRow)
-              .where("id", "=", latestRow.id)
-              .returningAll()
-              .executeTakeFirstOrThrow()
-              .catch((err) => {
-                throw parseStoreError(err, updateObject);
-              });
-            rows.push(row);
-          }
-
-          return rows;
-        });
+    if (typeof data === "function") {
+      const query = db
+        .withSchema(namespaceInfo.userNamespace)
+        .selectFrom(tableName)
+        .selectAll()
+        .where((eb) =>
+          buildWhereConditions({
+            eb,
+            where,
+            table,
+            encoding: kind,
+          }),
+        )
+        .orderBy("id", "asc");
+
+      const rows: UserRecord[] = [];
+      let cursor: DatabaseColumn = null;
+
+      while (true) {
+        const _rows = await db.wrap(
+          { method: `${tableName}.updateMany` },
+          async () => {
+            const latestRows: DatabaseRecord[] = await query
+              .limit(MAX_BATCH_SIZE)
+              .$if(cursor !== null, (qb) => qb.where("id", ">", cursor))
+              .execute();
+
+            const rows: DatabaseRecord[] = [];
+
+            for (const latestRow of latestRows) {
+              const current = decodeRow(latestRow, table, kind);
+              const updateObject = data({ current });
+              // Here, `latestRow` is already encoded, so we need to exclude it from `encodeRow`.
+              const updateRow = {
+                id: latestRow.id,
+                ...encodeRow(updateObject, table, kind),
+              };
+
+              const row = await db
+                .withSchema(namespaceInfo.userNamespace)
+                .updateTable(tableName)
+                .set(updateRow)
+                .where("id", "=", latestRow.id)
+                .returningAll()
+                .executeTakeFirstOrThrow()
+                .catch((err) => {
+                  throw parseStoreError(err, updateObject);
+                });
+              rows.push(row);
+            }
+
+            return rows.map((row) => decodeRow(row, table, kind));
+          },
+        );
 
-        return rows.map((row) => decodeRow(row, table, kind));
-      } else {
+        rows.push(..._rows);
+
+        if (_rows.length === 0) {
+          break;
+        } else {
+          cursor = encodeValue(_rows[_rows.length - 1].id, table.id, kind);
+        }
+      }
+
+      return rows;
+    } else {
+      return db.wrap({ method: `${tableName}.updateMany` }, async () => {
```

Comment:
Is it OK that the path above is async/await and this one returns the promise?

Reply:
yes